Add groups parameter to Convolutions

Michaeljurado24 opened the pull request:
**Motivation and context:**

Currently nengo does not support group convolution. Grouped convolutions are important because when groups == n_channels == n_filters, they are functionally equivalent to Depthwise Convolutions. (These in turn can be used to construct efficient separable convolutions which [greatly reduce computational overhead](https://towardsdatascience.com/a-basic-introduction-to-separable-convolutions-b99ec3102728) compared to regular convolutions. 

Furthermore adding grouped convolutions in nengo is a stepping stone to adding grouped convolutions in nengo-loihi. 

Issue 
https://github.com/nengo/nengo/issues/1675

Related Forum Posts
https://forum.nengo.ai/t/adding-a-groups-parameter-for-convolutions-in-nengo-loihi/2035/2
https://forum.nengo.ai/t/trying-to-understand-nengo-transforms-convolution/1844/4?u=mjurado3

**Interactions with other PRs:**
None

**How has this been tested?**
Primarily Through [this function](https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/builder/tests/test_transforms_conv.py#L18). I also manually compared the new convolutions against a tensorflow implementation of a convolution. 

**How long should this take to review?**
Quick since the only major change comes from [this function](https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/_vendor/npconv2d/conv2d.py#L184)

**Where should a reviewer start?**
1. The reviewer should probably first ask himself if it is worth accepting this pull request since TransposeConvolutions still do not currently accept a groups argument. 
2. If they deem that this is an acceptable, they should then move onto reviewing nengo/_vendor/npconv2d.py. 
3. Then should move onto the modified [Convolution Transform](https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/transforms.py#L351)

**Types of changes:**

- New-Feature: Low level backend convolution changes
- New-Feature: High Level addition of groups parameter to nengo.Convolution


**Checklist:**
<!--- Go over all the following points. Put an `x` in all the boxes that apply. -->
<!--- If a box is not applicable, please justify below the checklist. -->
<!--- If you're unsure about any of these, don't hesitate to ask. -->
<!--- We're here to help! -->

- [x] I have read the **CONTRIBUTING.rst** document.
- [x] I have updated the documentation accordingly.
- [x] I have included a changelog entry.
- [x] I have added tests to cover my changes.
- [x] I have run the test suite locally and all tests passed.

**Still to do:**
- [ ] Add groups param to nengo.ConvolutionTranspose
- [ ] Add groups param to this unit [test](https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/tests/test_transforms_conv.py#L53). (This involves completing the prior task first)

hunse commented on nengo/_vendor/npconv2d/conv2d.py:
```diff
@@ -185,21 +185,30 @@ def conv2d(x, w, pad='SAME', stride=(1, 1)):
-    y = y.reshape([xs[0], xs[1], xs[2], -1])
-    return y
+    x = x.reshape([groups, xs[1] * xs[2]* xs[3], -1])
+    y = np.einsum('ikj,ijm->kim', x, w)
```
I think it might be better to avoid the use of `einsum` in the original case (i.e. where `groups=1`), both for speed and clarity. I think the best approach might be to have a separate function that deals with the grouped case. Preserving the original function will mean we can both test against it, and have it for reference.

hunse commented on nengo/_vendor/npconv2d/conv2d.py:
```diff
@@ -185,21 +185,30 @@ def conv2d(x, w, pad='SAME', stride=(1, 1)):
     """
     ksize = w.shape[:2]
+    c = x.shape[-1]
+    groups = c // w.shape[2] # number of groups inferred
```
It might be good to assert that `c % w.shape[2] == 0`. If that isn't the case, my guess is that at best, the output will not be what the user expects (as they probably made a mistake in specifying the weight shape).

hunse commented on nengo/builder/tests/test_transforms_conv.py:
```diff
@@ -14,8 +14,9 @@
 @pytest.mark.parametrize("kernel0", (4, 5))
 @pytest.mark.parametrize("kernel1", (4, 5))
 @pytest.mark.parametrize("padding", ("same", "valid"))
+@pytest.mark.parametrize("groups",  (1, 8, 32))
```
There are already a lot of parametrizations for this test, and adding "groups" makes it go from 64 to 192. I know they're quite fast, but it adds clutter to the log and I'm not convinced it's necessary. Also, it might be good to test with different numbers of in and out channels, since that's directly related to groups.

The way I would do this is to take the body of this function, and make a `_test_convinc_2d` helper function. Then you can have two functions, `test_convinc_2d` and `test_grouped_convinc_2d`, that test the ungrouped and grouped cases. Since we're splitting up `conv2d` into grouped and ungrouped versions, this will also allow testing each one separately.

hunse commented:
I don't think it's necessary to also add groups for `ConvolutionTranspose` at this time, particularly if it's not something that you need.

Michaeljurado24 commented on nengo/tests/test_reprs.py:
```diff
@@ -576,13 +576,15 @@ def test_transforms():
     check_repr(Convolution(n_filters=3, input_shape=(3, 4, 5), channels_last=False))
     check_repr(Convolution(n_filters=3, input_shape=(3, 4, 5), strides=(2, 2)))
     check_repr(Convolution(n_filters=3, input_shape=(3, 4, 5), padding="same"))
+    check_repr(Convolution(n_filters=3, input_shape=(3, 4, 5), groups = 3))
```
One thing I notice is that when I run `pre-commit run --all-files` all of the changes I made get reformatted. However, a ton of other files get changed as well as in the example below. 
![image](https://user-images.githubusercontent.com/18356610/154752120-c71879cd-2adc-4f64-af83-5e4f71444933.png)

Is this okay?

tbekolay replied:
This is due to the `pre-commit` hook being out of date the latest version of Black; if you don't mind, can you remove the hook with `pre-commit uninstall`, then in your usual environment do `pip install black` and then `black .`? That should remove those erroneous changes.

Michaeljurado24 replied:
thank you

Michaeljurado24 commented:
In the new commit I addressed these changes and added a few unit tests to specifically test for grouped convolutions. Here they are

1. https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/builder/tests/test_transforms_conv.py#L95
2. https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/tests/test_transforms_conv.py#L52
3. https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/tests/test_transforms_conv.py#L199

hunse commented on nengo/_vendor/npconv2d/conv2d.py:
```diff
@@ -202,6 +202,39 @@ def conv2d(x, w, pad='SAME', stride=(1, 1)):
+    c = x.shape[-1]
+
+    # infer number of groups
+    assert c % w.shape[2] == 0 
```
Add a message like `f"Number of channels ({c}) must be divisible by number of groups ({w.shape[2]})"`

hunse replied:
Actually, I guess `w.shape[2]` isn't the number of groups, it's the channels divided by groups. I'm not quite sure what to call it...

hunse replied:
Maybe "channels per group"? Thus e.g. `f"Number of channels ({c}) must be divisible by channels-per-group ({w.shape[2]})"`

hunse commented on nengo/_vendor/npconv2d/conv2d.py:
```diff
@@ -202,6 +202,39 @@ def conv2d(x, w, pad='SAME', stride=(1, 1)):
+    Returns:
+        y: [N, H', W', K]
+    """
+    ksize = w.shape[:2]
```
While I'm thinking of it, we should probably assert that both `x` and `w` have `ndim == 4`, both here and in `conv2d`.

hunse commented on nengo/builder/tests/test_transforms_conv.py:
```diff
@@ -88,6 +88,93 @@ def test_convinc_2d(
+):
+    correlate2d = pytest.importorskip("scipy.signal").correlate2d
+
+    shape0, shape1 = 16, 17
```
What I was thinking for the test is that we could have one `_test_convinc_2d` helper function, that is called by both `test_convinc_2d` and `test_grouped_convinc_2d`. That way, there's not all the duplicated code.

hunse commented on nengo/builder/tests/test_transforms_conv.py:
```diff
@@ -96,11 +183,9 @@ def test_convtransposeinc_2d(
     channels_last, strides, kernel_size, padding, rng, allclose, plt
 ):
     """Test ConvTransposeInc by ensuring it is the transpose of ConvInc.
-
```
I think these blank lines were mistakenly removed.

hunse commented on nengo/tests/test_transforms_conv.py:
```diff
@@ -44,6 +44,46 @@ def test_convolution_shape(padding, stride, k_size, x_mul, rng, allclose):
+@pytest.mark.parametrize("k_size", (1, 2))
+@pytest.mark.parametrize("stride", (1, 2))
+@pytest.mark.parametrize("padding", ("same", "valid"))
+def test_convolution_group_shape(padding, stride, k_size, x_mul, groups, rng, allclose):
```
Again, I would have a common helper function for both this and the test above, so that we're not duplicating code.

hunse commented on nengo/tests/test_transforms_conv.py:
```diff
@@ -151,6 +191,93 @@ def test_convolution(
+    rng,
+    seed,
+):
+    input_d = 4
```
Again, common helper function with the test above.

Michaeljurado24 commented:
In the latest commit I added a few helper functions as suggested and also added a check for ndim == 4. 

I should also mention that I added an extra helper method called [scipy_conv2d](https://github.com/Michaeljurado24/nengo/blob/Convolution_groups/nengo/builder/tests/test_transforms_conv.py#L11) and scipy_conv2d_grouped to reduce tab nesting.

hunse commented:
Thanks for the changes! I'll take a look now.

Can you please sign the CAA [here](https://www.nengo.ai/caa/)? Just add your name to the People page as it describes at the top to sign. You'll have to do that before I can merge this PR.

hunse commented:
I've made some fixups. Let me know if they look good, @Michaeljurado24. I'll squash everything together before merge.

I'll take one last look once the CAA is signed before I merge, but I think that this is looking good!

I've made the `convolution-groups-ci` branch, that's just a copy of this for running CI. The tests passed for me locally, but I'll need to check the CI tests and coverage before merge.

Michaeljurado24 commented:
Everything looks good! I made a pull request with my name added to the CAA

hunse commented:
I've squashed everything down, which I think should get the last CI test to pass (it was a problem with the commit message being too long).

Note that it'll still be at least another week before this gets merged, because there are others on the team who want to take a quick look first. That said, I think this is quite close to what will be the final version, so feel free to build off this branch for e.g. a NengoLoihi implementation.

Michaeljurado24 commented:
So for adding grouped convolutions to nengo-loihi, I need to first add them to nengo-dl. 
For adding grouped convolutions to nengo-dl, I am thinking that I only need to modify the  [ConvertConv function](https://github.com/nengo/nengo-dl/blob/4abc90fa7a020cb30a58edfd749815884661605e/nengo_dl/converter.py#L1320) and [a unit test](https://github.com/nengo/nengo-dl/blob/master/nengo_dl/tests/test_converter.py#L37). Are there any other areas you know of which would need to be edited for a pull request to be accepted?

hunse commented:
The most important thing to update is actually this class, [`ConvIncBuilder`](https://github.com/nengo/nengo-dl/blob/26b7511d7a9d4d273c70eae5f01fcd1e794387f3/nengo_dl/transform_builders.py#L53). This is the class that takes a Nengo `ConvInc` operator and produces the appropriate TensorFlow code. So implementing the `groups` parameter there and adding a unit test would be the first step.

The `ConvertConv` class is part of the Keras to Nengo converter, which is used to import Keras models into Nengo. It should also be updated, but hopefully that's not too much more involved than just passing the groups parameter from the Keras layer to the Nengo `Convolution` transform.

Michaeljurado24 commented:
Okay, I just made a pull request for adding grouped convolutions to nengo-dl. Surprisingly I did not need to edit ConvIncBuilder. We can move further discussion about this to the [other pull request.](https://github.com/nengo/nengo-dl/pull/223)

tbekolay commented:
Hi @Michaeljurado24, thanks for this contribution! It's been merged now as b6002e4035f4d7641d0c740c3a9c8fab9e4d4889.
